server/src/middleware/etag.rs

```diff
@@ -0,0 +1,154 @@
+                .map_into_right_body());
+            })
+        } else {
+            let fut = self.service.call(req);
+            Box::pin(async move {
+                let res: ServiceResponse<B> = fut.await?;
+                match method {
+                    Method::GET => {
+                        let mut modified = true;
+                        let mut payload: Option<Bytes> = None;
+                        let mut res = res.map_body(|_h, body| match body.size() {
+                            BodySize::Sized(_size) => {
+                                let bytes = body.try_into_bytes().unwrap_or_else(|_| Bytes::new());
+                                payload = Some(bytes.clone());
+
+                                bytes.clone().boxed()
+                            }
+                            _ => body.boxed(),
+                        });
+                        if let Some(bytes) = payload {
+                            let response_hash = xxh3_128(&bytes);
+                            let base64 = base64::prelude::BASE64_URL_SAFE
+                                .encode(response_hash.to_le_bytes());
+                            let mut buff = Buffer::new();
+                            let _ = write!(buff, "{:x}-{}", bytes.len(), base64);
+                            let tag = EntityTag::new_weak(buff.to_string());
+                            if let Some(request_etag_header) = request_etag_header {
+                                if request_etag_header == IfNoneMatch::Any
+                                    || request_etag_header.to_string() == tag.to_string()
+                                {
+                                    modified = false
+                                }
+                            }
+                            if modified {
+                                if let Ok((name, value)) = ETag(tag.clone()).try_into_pair() {
+                                    res.headers_mut().insert(name, value);
+                                }
+                            }
+                        }
+
+                        Ok(match modified {
+                            false => res
+                                .into_response(HttpResponse::NotModified().body(BodyNone::new()))
+                                .map_into_right_body(),
+                            true => res.map_into_left_body(),
+                        })
+                    }
+                    _ => Ok(res.map_into_boxed_body().map_into_left_body()),
+                }
+            })
+        }
+    }
+}
+
+fn header_to_edgetoken(header: Option<&HeaderValue>) -> Option<EdgeToken> {
+    header
+        .map(|h| h.to_str().unwrap())
+        .and_then(|header_str| EdgeToken::from_str(header_str).ok())
+}
+
+fn we_know_this_etag_from_upstream(
+    etag_cache: Arc<DashMap<EdgeToken, EntityTag>>,
+    client_token: &Option<EdgeToken>,
+    if_none_match: &Option<IfNoneMatch>,
+) -> bool {
+    match (if_none_match, client_token) {
```

Comment:
The match one looks odd to me here. You can do a tuple destructuring `if let Some` here:

fn we_know_this_etag_from_upstream(
    etag_cache: Arc<DashMap<EdgeToken, EntityTag>>,
    client_token: &Option<EdgeToken>,
    if_none_match: &Option<IfNoneMatch>,
) -> bool {
    if let (Some(if_none), Some(token)) = (if_none_match, client_token) {
        etag_cache.get(token).map_or(false, |etag| {
            if_none == &IfNoneMatch::Any || if_none.to_string() == etag.to_string()
        })
    } else {
        false
    }
}

Entirely optional